go/vt/vtgate/plan_execute.go

```diff
@@ -344,13 +349,30 @@ func (e *Executor) executePlan(
 
 // rollbackExecIfNeeded rollbacks the partial execution if earlier it was detected that it needs partial query execution to be rolled back.
 func (e *Executor) rollbackExecIfNeeded(ctx context.Context, safeSession *econtext.SafeSession, bindVars map[string]*querypb.BindVariable, logStats *logstats.LogStats, err error) error {
-	if safeSession.InTransaction() && safeSession.IsRollbackSet() {
+	if !safeSession.InTransaction() {
+		return err
+	}
+	if e.rollbackOnFatalTxError(ctx, safeSession, err) {
+		return err
+	}
+
+	if safeSession.IsRollbackSet() {
 		rErr := e.rollbackPartialExec(ctx, safeSession, bindVars, logStats)
```

Comment:
StreamExecute calls this only when it is called for DML query. This is valid for Select query as well inside a transaction

```
		// Check if there was partial DML execution. If so, rollback the effect of the partially executed query.
		if err != nil {
			if !canReturnRows(plan.Type) {
				return e.rollbackExecIfNeeded(ctx, safeSession, bindVars, logStats, err)
			}
			return err
		}
```